[TDX] Added Intel TDX support to helm charts

JakubLedworowski opened the pull request:
## Description

Confidential computing in AI in the cloud focuses on protecting sensitive data and computations from unauthorized access and tampering. It uses advanced security technologies, such as hardware-based isolation and encryption, to create secure environments where data and AI models can be processed safely. This ensures that even cloud service providers cannot access the data, providing a higher level of privacy and security. By leveraging confidential computing, organizations can confidently use AI in the cloud for tasks that involve sensitive information, such as healthcare data analysis or financial predictions, while complying with strict data protection regulations.

Subcharts enabled to be run with Intel TDX:
 - embedding-usvc
 - guardrails-usvc
 - llm-uservice
 - redis-vector-db
 - reranking-usvc
 - retriever-usvc
 - tei
 - teirerank
 - tgi
 - vllm
- ui
- data-prep
- nginx

## Issues

List the issue or RFC link this PR is working on. If there is no such link, please mark it as `n/a`.

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

List the newly introduced 3rd party dependency if exists.

## Tests

Tested the generated templates with and without Intel TDX.
Tested the deployment of `chatqna` example on the running cluster.

JakubLedworowski commented on helm-charts/TDX.md:
```diff
@@ -0,0 +1,100 @@
+   echo "runtimeRequestTimeout: 30m" | sudo tee -a /etc/kubernetes/kubelet-config.yaml > /dev/null 2>&1
+   # Vanilla Kubernetes installation
+   sudo sed -i 's/runtimeRequestTimeout: .*/runtimeRequestTimeout: 30m/' /var/lib/kubelet/config.yaml > /dev/null 2>&1
+   # Wait for the node to be ready
```
restart and wait

JakubLedworowski commented on helm-charts/TDX.md:
````diff
@@ -0,0 +1,100 @@
+      --set vllm.tdxEnabled=true --set vllm.resources.limits.memory=8Gi
+   ```
+
+   > [!NOTE]
````
fix the NOTE rendering

lianhao commented on helm-charts/common/common-library/Chart.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: v2
+name: common-library
+description: A Helm chart for the common templates
+version: 1.2.0
```
Please use '0-latest' as the default version. In each release, we'll replace the version value automatically by CI.

dcmiddle replied:
Please include this rationale in the commit message

lianhao reviewed (changes requested):
We should combine the common-library and tdx charts into a single tdx library chart. The benefit of using library chart can be found [at here](https://helm.sh/docs/topics/library_charts/#library-chart-benefits)

JakubLedworowski commented on helm-charts/TDX.md:
````diff
@@ -0,0 +1,100 @@
+3. Depending on the location of your kubelet config file, increase the kubelet timeout and wait until the node is `Ready`:
+
+   ```bash
+   # Kubespray installation
````
They provide different ways of setting up cluster
https://github.com/opea-project/GenAIInfra/blob/main/README.md?plain=1#L26

JakubLedworowski commented on helm-charts/TDX.md:
```diff
@@ -0,0 +1,100 @@
+      --set retriever-usvc.tdxEnabled=true --set retriever-usvc.resources.limits.memory=8Gi \
+      --set tei.tdxEnabled=true --set tei.resources.limits.memory=8Gi \
+      --set teirerank.tdxEnabled=true --set teirerank.resources.limits.memory=8Gi \
+      --set vllm.tdxEnabled=true --set vllm.resources.limits.memory=8Gi
```
This is already written in the helm docs so I won't repeat it here
https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/README.md?plain=1#L81

ksandowi commented on helm-charts/TDX.md:
````diff
@@ -0,0 +1,100 @@
+2. Deploy the Helm Chart setting the `tdxEnabled` flag for each microservice you want to run using Intel TDX, for example:
+
+   ```
+   helm install $myrelease $chartname \
````
helm install $myrelease $chartname \
   --set global.HUGGINGFACEHUB_API_TOKEN="${HFTOKEN}" --set vllm.LLM_MODEL_ID="${MODELNAME}" \
   --set nginx.tdxEnabled=true \
   --set chatqna-ui.tdxEnabled=true --set chatqna-ui.resources.limits.memory=2Gi \
   --set data-prep.tdxEnabled=true --set data-prep.resources.limits.memory=11Gi \
   --set redis-vector-db.tdxEnabled=true --set redis-vector-db.resources.limits.memory=4Gi \
   --set retriever-usvc.tdxEnabled=true --set retriever-usvc.resources.limits.memory=7Gi \
   --set tei.tdxEnabled=true --set tei.resources.limits.memory=4Gi \
   --set teirerank.tdxEnabled=true --set teirerank.resources.limits.memory=6Gi \
   --set vllm.tdxEnabled=true --set vllm.resources.limits.memory=70Gi

ksandowi reviewed (approved):
Update helm install with following minimal resources required for TDX

JakubLedworowski commented on helm-charts/common/tei/templates/deployment.yaml:
```diff
@@ -17,13 +17,11 @@ spec:
       labels:
         {{- include "tei.selectorLabels" . | nindent 8 }}
     spec:
+      {{- include "common.runtimeClassName" . | nindent 6 }}
```
done, thanks!

JakubLedworowski commented:
> We should combine the common-library and tdx charts into a single tdx library chart. The benefit of using library chart can be found [at here](https://helm.sh/docs/topics/library_charts/#library-chart-benefits)

Hi @lianhao, I tried to do it with a single library at the beginning, but the problem I faced is that there is a common logic for including podAnnotations. With annotations for Intel TDX we have to merge with the existing solution. That is why I introduced a common-library, which is always included and inside it checks if tdx values are present.

If you have an idea how to make it a single library I am open to suggestions :)

JakubLedworowski commented:
> > We should combine the common-library and tdx charts into a single tdx library chart. The benefit of using library chart can be found [at here](https://helm.sh/docs/topics/library_charts/#library-chart-benefits)
> 
> Hi @lianhao, I tried to do it with a single library at the beginning, but the problem I faced is that there is a common logic for including podAnnotations. With annotations for Intel TDX we have to merge with the existing solution. That is why I introduced a common-library, which is always included and inside it checks if tdx values are present.
> 
> If you have an idea how to make it a single library I am open to suggestions :)

After a call we simplified the solution into a single library included as permanent dependency.
A common part with annotations is always included but the tdx part is only appended when tdxEnabled is true.

lianhao reviewed (approved):
Please fix the DCO issue. Others seems fine to me

lianhao commented:
CI failed due to 2 known issue:
- opea-project/GenAIComps#1293
- opea-project/GenAIComps#1295

eero-t reviewed (commented):
Something to fix in future PR, is removal of now redundant & misleading variable declarations from values files (run from `helm-charts/` dir):
```
$ varname=podAnnotations; \
  echo "Value files with no matching '$varname' variable use in templates:"; \
  for i in $(grep -l $varname */values.yaml common/*/*values.yaml); do \
    if ! grep -q $varname ${i%/*}/templates/*.yaml; then \
      echo "- $(grep -H $varname $i)"; \
    fi; \
  done
Value files with no matching 'podAnnotations' variable use in templates:
- common/data-prep/values.yaml:podAnnotations: {}
- common/embedding-usvc/values.yaml:podAnnotations: {}
- common/guardrails-usvc/values.yaml:podAnnotations: {}
- common/llm-uservice/values.yaml:podAnnotations: {}
- common/nginx/values.yaml:podAnnotations: {}
- common/redis-vector-db/values.yaml:podAnnotations: {}
- common/reranking-usvc/values.yaml:podAnnotations: {}
- common/retriever-usvc/values.yaml:podAnnotations: {}
- common/teirerank/values.yaml:podAnnotations: {}
- common/tei/values.yaml:podAnnotations: {}
- common/tgi/values.yaml:podAnnotations: {}
- common/ui/values.yaml:podAnnotations: {}
- common/vllm/values.yaml:podAnnotations: {}
```

eero-t commented:
> Hello @eero-t, the podAnnotations are still used like before but in commonlib: https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/common/commonlib/templates/_helpers.tpl#L3
> 
> The TDX-specific annotations are merged with the podAnnotations is they are passed by the user.

I see, thanks!

(I haven't used library charts myself, but they seem really useful.  They complicate quite bit checking values <-> template variable references though. :-/)
